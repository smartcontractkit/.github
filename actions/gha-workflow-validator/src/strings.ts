export const VALIDATOR_IGNORE_LINE = "ghv-ignore!";
export const FIXING_ERRORS = `


This CI step validates Github Actions workflows. If invoked by a pull request, it will only operate on the changed files.

It currently validates: Actions References, Actions Runner Types

### Fixing Errors

<details>
<summary>Action References (sha-ref, version-comment, node-version) </summary>

This validation is required to ensure that the action references use immutable SHAs, have a version comment, and are not using node16 or earlier.

The proper format for referencing a Github Action external to the repository is as follows:

\`<owner>/<repo>/<optional path>@<commit SHA> # <version tag relating to the SHA>\`

<details>
<summary>Examples</summary>

\`\`\`
actions/cache@ab5e6d0c87105b4c9c2047343972218f562e4319 # v4.0.1

smartcontractkit/chainlink-github-actions/github-app-token-issuer@5874ff7211cf5a5a2670bb010fbff914eaaae138 # v2.3.12
\`\`\`
</details>

##### <ref> is not a valid SHA

* Please reference a specific commit. This is because tags are mutable and pose a security risk
* Do not use things like \`@main\`, \`@branch/feature\`, \`@v4\`, or \`@v4.0.0\`.

##### No version comment found

* Ensure you have left a comment indicating the git tag (or version) associated with the SHA reference.
* \`@<commit> # v4.0.0\`

##### Action is using node...

* The action added is supposed to be run using a version that is not \`node20\`. This might create issues due to Github deprecating actions using \`node16\` and earlier.

</details>

<details>
<summary>Actions Runners (runners)</summary>

This validation is required to limit the cost of high cost runners. See [Github Actions Runner Guidance](https://smartcontract-it.atlassian.net/l/cp/Rw0Gc08x).

##### Actions runner is to expensive

* Consider using a smaller runner type. This will help reduce the cost of the Github Actions workflow.
* If using macOs runners, consider using ubuntu runners instead.
* If you must use an upgraded runner then see Ignoring Errors section below.

</details>

### Ignoring Errors

You can use the following string to ignore a line from validation \`${VALIDATOR_IGNORE_LINE}\` (must be inlined).

When adding this it will trigger an error, this is expected. Please reach out to \`#team-releng\` for approval and bypass.

<details>
<summary>Examples</summary>

\`\`\`
runs-on: ubuntu-latest-99cores-999GB # ${VALIDATOR_IGNORE_LINE}

uses: org/action@sha # v1.0.0-node16 ${VALIDATOR_IGNORE_LINE}
\`\`\`
</details>
`;

export function htmlLink(text: string, url: string) {
  return `<a href="${url}">${text}</a>`;
}
