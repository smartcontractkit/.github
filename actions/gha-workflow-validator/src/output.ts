import * as core from "@actions/core";
import {
  SummaryTableRow as TRow,
  SummaryTableCell as TCell,
} from "@actions/core/lib/summary.js";
import {
  FileValidationResult,
  LineValidationResult,
  ValidationType,
} from "./validations/validation-check.js";
import { FIXING_ERRORS, htmlLink } from "./strings";

export function logErrors(
  validationResults: FileValidationResult[],
  annotatePR: boolean = false,
) {
  for (const fileResults of validationResults) {
    for (const lineResults of fileResults.lineValidations) {
      if (lineResults.messages.length === 0) {
        continue;
      }

      const validationMessages = lineResults.messages.map(
        (error, index) =>
          `  ${index + 1}. ${error.message} (${error.type} / ${error.severity})`,
      );

      const maxSeverity = getMaxSeverityForLine(lineResults);
      const logLine = `file: ${fileResults.filename} @ line: ${lineResults.line.lineNumber} - ${validationMessages.join(",")}`;

      if (maxSeverity === "ignored") {
        core.info(`(ignored) ${logLine}`);
        continue;
      }

      const loggingMethod =
        maxSeverity === "warning" ? core.warning : core.error;
      loggingMethod(logLine);
      if (annotatePR) {
        loggingMethod(validationMessages.join("\n"), {
          file: fileResults.filename,
          startLine: lineResults.line.lineNumber,
        });
      }
    }
  }
}

export async function setSummary(
  validationResults: FileValidationResult[],
  fileUrlPrefix: string,
) {
  // Flatten the validation results into a list of rows for the summary table.
  // For each cell spanning multiple rows, you need to include it in the first row it appears in, with the rest of the cells for the first row
  // of it's rowspan. Then for each subsequent row that it spans, you only include only the other cells in that row.
  // We have filename and line numbers which span multiple rows.
  // Below is an example of the array entries in the errorRows array. Note: empty cells are not included in each array row
  // Filename LineNumber Level Type Message
  // fileA   1           error foo  bar
  //                     error foo  bar
  //         2           error foo  bar
  //                     error foo  bar
  // fileB   1           error foo  bar
  //                     error foo  bar
  //                     error foo  bar

  const headerRow: TRow = [
    { data: "Filename", header: true },
    { data: "Line Number", header: true },
    { data: "Level", header: true },
    { data: "Type", header: true },
    { data: "Message", header: true },
  ];

  const errorRows = validationResults.flatMap((fileResult) => {
    const filename = fileResult.filename;
    const lineErrorRows: TRow[] = [];

    fileResult.lineValidations.forEach((lineValidation) => {
      if (lineValidation.messages.length === 0) return;
      const ignoreErrors = lineValidation.line.ignored;

      const lineNumberCell: TCell = {
        data: htmlLink(
          `${lineValidation.line.lineNumber}`,
          `${fileUrlPrefix}/${filename}#L${lineValidation.line.lineNumber}`,
        ),
        rowspan: `${lineValidation.messages.length}`,
      };

      const validationRows: TRow[] = lineValidation.messages
        .filter(
          (message) =>
            message.type === ValidationType.IGNORE_COMMENT || !ignoreErrors,
        )
        .map((message, index) => {
          const levelCell: TCell = { data: message.severity };
          const typeCell: TCell = { data: message.type };
          const messageCell: TCell = { data: message.message };

          // First row includes the line number, subsequent rows do not
          return index === 0
            ? [lineNumberCell, levelCell, typeCell, messageCell]
            : [levelCell, typeCell, messageCell];
        });

      lineErrorRows.push(...validationRows);
    });

    if (lineErrorRows.length === 0) return [];

    const filenameCell: TCell = {
      data: filename,
      rowspan: `${lineErrorRows.length}`,
    };

    // Insert the filename in the first row and return all rows for the file
    return lineErrorRows.map((row, index) =>
      index === 0 ? [filenameCell, ...row] : row,
    );
  });

  await core.summary
    .addTable([headerRow, ...errorRows])
    .addSeparator()
    .addRaw(FIXING_ERRORS)
    .write();
}

function getMaxSeverityForLine(lvr: LineValidationResult): string {
  const maxSeverity = lvr.messages.reduce((acc, curr) => {
    if (acc === "error" || curr.severity === "error") return "error";
    if (acc === "warning" || curr.severity === "warning") return "warning";
    if (acc === "ignored" || curr.severity === "ignored") return "ignored";
    return "";
  }, "");

  if (maxSeverity === "") {
    core.warning(
      `Invalid validation severity found. Defaulting to "error" for ${lvr.filename} @ ${lvr.line.lineNumber}`,
    );
    return "error";
  }

  return maxSeverity;
}
