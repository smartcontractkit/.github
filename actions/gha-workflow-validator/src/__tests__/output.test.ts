import { setSummary } from "../output.js";
import { vi, afterEach, describe, it, expect } from "vitest";
import * as core from "@actions/core";
import {
  FileValidationResult,
  ValidationType,
} from "../validations/validation-check.js";
import { FIXING_ERRORS } from "../strings.js";

vi.mock("@actions/core", async (importOriginal) => {
  const stubs = (await import("./__helpers__/test-utils.js")).coreLoggingStubs();

  const summary: any = {};
  summary.addTable = vi.fn().mockReturnValue(summary);
  summary.addSeparator = vi.fn().mockReturnValue(summary);
  summary.addRaw = vi.fn().mockReturnValue(summary);
  summary.write = vi.fn().mockReturnValue(summary);

  return {
    ...stubs,
    summary,
  }
});

describe(setSummary.name, () => {

  afterEach(() => {
    vi.clearAllMocks();
  });

  it("should set summary with no input", async () => {
    setSummary([], "prefix.com");
    const { summary } = core;

    expect(summary.addTable).toHaveBeenCalledTimes(1);
    expect(summary.addTable).toMatchSnapshot();
    expect(summary.addSeparator).toHaveBeenCalledTimes(1);
    expect(summary.addRaw).toHaveBeenCalledTimes(1);
    expect(summary.addRaw).toHaveBeenCalledWith(FIXING_ERRORS);
    expect(summary.write).toHaveBeenCalledTimes(1);
  });

  it("should set summary with no errors ", async () => {
    const validationResult: FileValidationResult[] = [
      {
        filename: "foo.yml",
        lineValidations: [
          {
            filename: "foo.yml",
            line: {
              lineNumber: 1,
              content: "line 1",
              operation: "add",
              ignored: false,
            },
            messages: []
          },
          {
            filename: "foo.yml",
            line: {
              lineNumber: 3,
              content: "line 2",
              operation: "add",
              ignored: false,
            },
            messages: []
          },
        ],
      }
    ]

    setSummary(validationResult, "prefix.com");
    const { summary } = core;

    expect(summary.addTable).toHaveBeenCalledTimes(1);
    expect(summary.addTable).toMatchSnapshot();
    expect(summary.addSeparator).toHaveBeenCalledTimes(1);
    expect(summary.addRaw).toHaveBeenCalledTimes(1);
    expect(summary.addRaw).toHaveBeenCalledWith(FIXING_ERRORS);
    expect(summary.write).toHaveBeenCalledTimes(1);
  });

  it("should set summary with single file", async () => {
    const validationResult: FileValidationResult[] = [
      {
        filename: "foo.yml",
        lineValidations: [
          {
            filename: "foo.yml",
            line: {
              lineNumber: 1,
              content: "line 1",
              operation: "add",
              ignored: false,
            },
            messages: [{
              type: ValidationType.SHA_REF,
              severity: "error",
              message: "sha ref error",
            }]
          },
          {
            filename: "foo.yml",
            line: {
              lineNumber: 3,
              content: "line 2",
              operation: "add",
              ignored: false,
            },
            messages: [{
              type: ValidationType.VERSION_COMMENT,
              severity: "warning",
              message: "version comment warning",
            }]
          },
        ],
      }
    ]

    setSummary(validationResult, "prefix.com");
    const { summary } = core;

    expect(summary.addTable).toHaveBeenCalledTimes(1);
    expect(summary.addTable).toMatchSnapshot();
    expect(summary.addSeparator).toHaveBeenCalledTimes(1);
    expect(summary.addRaw).toHaveBeenCalledTimes(1);
    expect(summary.addRaw).toHaveBeenCalledWith(FIXING_ERRORS);
    expect(summary.write).toHaveBeenCalledTimes(1);
  });

  it("should set summary with multiple files", async () => {
    const validationResult: FileValidationResult[] = [
      {
        filename: "foo.yml",
        lineValidations: [
          {
            filename: "foo.yml",
            line: {
              lineNumber: 1,
              content: "line 1",
              operation: "add",
              ignored: false,
            },
            messages: [{
              type: ValidationType.SHA_REF,
              severity: "error",
              message: "sha ref error",
            }]
          },
          {
            filename: "foo.yml",
            line: {
              lineNumber: 3,
              content: "line 2",
              operation: "add",
              ignored: false,
            },
            messages: [{
              type: ValidationType.VERSION_COMMENT,
              severity: "warning",
              message: "version comment warning",
            }]
          },
        ],
      },
      {
        filename: "bar.yml",
        lineValidations: [
          {
            filename: "foo.yml",
            line: {
              lineNumber: 1337,
              content: "line 1337",
              operation: "add",
              ignored: false,
            },
            messages: [{
              type: ValidationType.SHA_REF,
              severity: "error",
              message: "sha ref error",
            }]
          },
          {
            filename: "foo.yml",
            line: {
              lineNumber: 1338,
              content: "line 1338",
              operation: "add",
              ignored: false,
            },
            messages: [{
              type: ValidationType.VERSION_COMMENT,
              severity: "warning",
              message: "version comment warning",
            }]
          },
        ],
      }
    ]

    setSummary(validationResult, "prefix.com");
    const { summary } = core;

    expect(summary.addTable).toHaveBeenCalledTimes(1);
    expect(summary.addTable).toMatchSnapshot();
    expect(summary.addSeparator).toHaveBeenCalledTimes(1);
    expect(summary.addRaw).toHaveBeenCalledTimes(1);
    expect(summary.addRaw).toHaveBeenCalledWith(FIXING_ERRORS);
    expect(summary.write).toHaveBeenCalledTimes(1);
  });
});
