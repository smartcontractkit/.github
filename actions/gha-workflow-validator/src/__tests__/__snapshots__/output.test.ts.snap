// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`setSummary > should set summary with multiple files 1`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        [
          {
            "data": "Filename",
            "header": true,
          },
          {
            "data": "Line Number",
            "header": true,
          },
          {
            "data": "Level",
            "header": true,
          },
          {
            "data": "Type",
            "header": true,
          },
          {
            "data": "Message",
            "header": true,
          },
        ],
        [
          {
            "data": "foo.yml",
            "rowspan": "2",
          },
          {
            "data": "<a href="prefix.com/foo.yml#L1">1</a>",
            "rowspan": "1",
          },
          {
            "data": "error",
          },
          {
            "data": "sha-ref",
          },
          {
            "data": "sha ref error",
          },
        ],
        [
          {
            "data": "<a href="prefix.com/foo.yml#L3">3</a>",
            "rowspan": "1",
          },
          {
            "data": "warning",
          },
          {
            "data": "version-comment",
          },
          {
            "data": "version comment warning",
          },
        ],
        [
          {
            "data": "bar.yml",
            "rowspan": "2",
          },
          {
            "data": "<a href="prefix.com/bar.yml#L1337">1337</a>",
            "rowspan": "1",
          },
          {
            "data": "error",
          },
          {
            "data": "sha-ref",
          },
          {
            "data": "sha ref error",
          },
        ],
        [
          {
            "data": "<a href="prefix.com/bar.yml#L1338">1338</a>",
            "rowspan": "1",
          },
          {
            "data": "warning",
          },
          {
            "data": "version-comment",
          },
          {
            "data": "version comment warning",
          },
        ],
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": {
        "addRaw": [MockFunction spy] {
          "calls": [
            [
              "


This CI step validates Github Actions workflows. If invoked by a pull request, it will only operate on the changed files.

It currently validates: Actions References, Actions Runner Types

### Fixing Errors

<details>
<summary>Action References (sha-ref, version-comment, node-version) </summary>

This validation is required to ensure that the action references use immutable SHAs, have a version comment, and are not using node16 or earlier.

The proper format for referencing a Github Action external to the repository is as follows:

\`<owner>/<repo>/<optional path>@<commit SHA> # <version tag relating to the SHA>\`

<details>
<summary>Examples</summary>

\`\`\`
actions/cache@ab5e6d0c87105b4c9c2047343972218f562e4319 # v4.0.1

smartcontractkit/chainlink-github-actions/github-app-token-issuer@5874ff7211cf5a5a2670bb010fbff914eaaae138 # v2.3.12
\`\`\`
</details>

##### <ref> is not a valid SHA

* Please reference a specific commit. This is because tags are mutable and pose a security risk
* Do not use things like \`@main\`, \`@branch/feature\`, \`@v4\`, or \`@v4.0.0\`.

##### No version comment found

* Ensure you have left a comment indicating the git tag (or version) associated with the SHA reference.
* \`@<commit> # v4.0.0\`

##### Action is using node...

* The action added is supposed to be run using a version that is not \`node20\`. This might create issues due to Github deprecating actions using \`node16\` and earlier.

</details>

<details>
<summary>Actions Runners (runners)</summary>

This validation is required to limit the cost of high cost runners. See [Github Actions Runner Guidance](https://smartcontract-it.atlassian.net/l/cp/Rw0Gc08x).
For specific runner costs see "[What are the per-minute costs of the runners?](https://smartcontract-it.atlassian.net/wiki/spaces/RE/pages/861241466/Github+Actions+Runner+Guidance#What-are-the-per-minute-costs-of-the-runners%3F)"

##### \`runner-macos\`

* MacOS runners are very expensive in comparison to Ubuntu runners. If you need ARM64 architecture, consider using the new ARM64-based Ubuntu runners.
* If you need a MacOS runner, consider using the base runner especially if this is a public repository.
* If you must use an upgraded runner then see Ignoring Errors section below.

##### \`runner-ubuntu\`

* Per-minute compute costs scale with the number of cores. The base Ubuntu runner is the most cost-effective, especially for public repositories.
* If you must use an upgraded runner then see Ignoring Errors section below.

</details>

### Ignoring Errors

You can use the following string to ignore a line from validation \`ghv-ignore!\` (must be inlined).

When adding this it will trigger an error, this is expected. Please reach out to \`#team-releng\` for approval and bypass.

<details>
<summary>Examples</summary>

\`\`\`
runs-on: ubuntu-latest-99cores-999GB # ghv-ignore!

uses: org/action@sha # v1.0.0-node16 ghv-ignore!
\`\`\`
</details>
",
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "addSeparator": [MockFunction spy] {
          "calls": [
            [],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "addTable": [MockFunction spy] {
          "calls": [
            [
              [
                [
                  {
                    "data": "Filename",
                    "header": true,
                  },
                  {
                    "data": "Line Number",
                    "header": true,
                  },
                  {
                    "data": "Level",
                    "header": true,
                  },
                  {
                    "data": "Type",
                    "header": true,
                  },
                  {
                    "data": "Message",
                    "header": true,
                  },
                ],
                [
                  {
                    "data": "foo.yml",
                    "rowspan": "2",
                  },
                  {
                    "data": "<a href="prefix.com/foo.yml#L1">1</a>",
                    "rowspan": "1",
                  },
                  {
                    "data": "error",
                  },
                  {
                    "data": "sha-ref",
                  },
                  {
                    "data": "sha ref error",
                  },
                ],
                [
                  {
                    "data": "<a href="prefix.com/foo.yml#L3">3</a>",
                    "rowspan": "1",
                  },
                  {
                    "data": "warning",
                  },
                  {
                    "data": "version-comment",
                  },
                  {
                    "data": "version comment warning",
                  },
                ],
                [
                  {
                    "data": "bar.yml",
                    "rowspan": "2",
                  },
                  {
                    "data": "<a href="prefix.com/bar.yml#L1337">1337</a>",
                    "rowspan": "1",
                  },
                  {
                    "data": "error",
                  },
                  {
                    "data": "sha-ref",
                  },
                  {
                    "data": "sha ref error",
                  },
                ],
                [
                  {
                    "data": "<a href="prefix.com/bar.yml#L1338">1338</a>",
                    "rowspan": "1",
                  },
                  {
                    "data": "warning",
                  },
                  {
                    "data": "version-comment",
                  },
                  {
                    "data": "version comment warning",
                  },
                ],
              ],
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "write": [MockFunction spy] {
          "calls": [
            [],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`setSummary > should set summary with no errors  1`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        [
          {
            "data": "Filename",
            "header": true,
          },
          {
            "data": "Line Number",
            "header": true,
          },
          {
            "data": "Level",
            "header": true,
          },
          {
            "data": "Type",
            "header": true,
          },
          {
            "data": "Message",
            "header": true,
          },
        ],
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": {
        "addRaw": [MockFunction spy] {
          "calls": [
            [
              "


This CI step validates Github Actions workflows. If invoked by a pull request, it will only operate on the changed files.

It currently validates: Actions References, Actions Runner Types

### Fixing Errors

<details>
<summary>Action References (sha-ref, version-comment, node-version) </summary>

This validation is required to ensure that the action references use immutable SHAs, have a version comment, and are not using node16 or earlier.

The proper format for referencing a Github Action external to the repository is as follows:

\`<owner>/<repo>/<optional path>@<commit SHA> # <version tag relating to the SHA>\`

<details>
<summary>Examples</summary>

\`\`\`
actions/cache@ab5e6d0c87105b4c9c2047343972218f562e4319 # v4.0.1

smartcontractkit/chainlink-github-actions/github-app-token-issuer@5874ff7211cf5a5a2670bb010fbff914eaaae138 # v2.3.12
\`\`\`
</details>

##### <ref> is not a valid SHA

* Please reference a specific commit. This is because tags are mutable and pose a security risk
* Do not use things like \`@main\`, \`@branch/feature\`, \`@v4\`, or \`@v4.0.0\`.

##### No version comment found

* Ensure you have left a comment indicating the git tag (or version) associated with the SHA reference.
* \`@<commit> # v4.0.0\`

##### Action is using node...

* The action added is supposed to be run using a version that is not \`node20\`. This might create issues due to Github deprecating actions using \`node16\` and earlier.

</details>

<details>
<summary>Actions Runners (runners)</summary>

This validation is required to limit the cost of high cost runners. See [Github Actions Runner Guidance](https://smartcontract-it.atlassian.net/l/cp/Rw0Gc08x).
For specific runner costs see "[What are the per-minute costs of the runners?](https://smartcontract-it.atlassian.net/wiki/spaces/RE/pages/861241466/Github+Actions+Runner+Guidance#What-are-the-per-minute-costs-of-the-runners%3F)"

##### \`runner-macos\`

* MacOS runners are very expensive in comparison to Ubuntu runners. If you need ARM64 architecture, consider using the new ARM64-based Ubuntu runners.
* If you need a MacOS runner, consider using the base runner especially if this is a public repository.
* If you must use an upgraded runner then see Ignoring Errors section below.

##### \`runner-ubuntu\`

* Per-minute compute costs scale with the number of cores. The base Ubuntu runner is the most cost-effective, especially for public repositories.
* If you must use an upgraded runner then see Ignoring Errors section below.

</details>

### Ignoring Errors

You can use the following string to ignore a line from validation \`ghv-ignore!\` (must be inlined).

When adding this it will trigger an error, this is expected. Please reach out to \`#team-releng\` for approval and bypass.

<details>
<summary>Examples</summary>

\`\`\`
runs-on: ubuntu-latest-99cores-999GB # ghv-ignore!

uses: org/action@sha # v1.0.0-node16 ghv-ignore!
\`\`\`
</details>
",
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "addSeparator": [MockFunction spy] {
          "calls": [
            [],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "addTable": [MockFunction spy] {
          "calls": [
            [
              [
                [
                  {
                    "data": "Filename",
                    "header": true,
                  },
                  {
                    "data": "Line Number",
                    "header": true,
                  },
                  {
                    "data": "Level",
                    "header": true,
                  },
                  {
                    "data": "Type",
                    "header": true,
                  },
                  {
                    "data": "Message",
                    "header": true,
                  },
                ],
              ],
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "write": [MockFunction spy] {
          "calls": [
            [],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`setSummary > should set summary with no input 1`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        [
          {
            "data": "Filename",
            "header": true,
          },
          {
            "data": "Line Number",
            "header": true,
          },
          {
            "data": "Level",
            "header": true,
          },
          {
            "data": "Type",
            "header": true,
          },
          {
            "data": "Message",
            "header": true,
          },
        ],
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": {
        "addRaw": [MockFunction spy] {
          "calls": [
            [
              "


This CI step validates Github Actions workflows. If invoked by a pull request, it will only operate on the changed files.

It currently validates: Actions References, Actions Runner Types

### Fixing Errors

<details>
<summary>Action References (sha-ref, version-comment, node-version) </summary>

This validation is required to ensure that the action references use immutable SHAs, have a version comment, and are not using node16 or earlier.

The proper format for referencing a Github Action external to the repository is as follows:

\`<owner>/<repo>/<optional path>@<commit SHA> # <version tag relating to the SHA>\`

<details>
<summary>Examples</summary>

\`\`\`
actions/cache@ab5e6d0c87105b4c9c2047343972218f562e4319 # v4.0.1

smartcontractkit/chainlink-github-actions/github-app-token-issuer@5874ff7211cf5a5a2670bb010fbff914eaaae138 # v2.3.12
\`\`\`
</details>

##### <ref> is not a valid SHA

* Please reference a specific commit. This is because tags are mutable and pose a security risk
* Do not use things like \`@main\`, \`@branch/feature\`, \`@v4\`, or \`@v4.0.0\`.

##### No version comment found

* Ensure you have left a comment indicating the git tag (or version) associated with the SHA reference.
* \`@<commit> # v4.0.0\`

##### Action is using node...

* The action added is supposed to be run using a version that is not \`node20\`. This might create issues due to Github deprecating actions using \`node16\` and earlier.

</details>

<details>
<summary>Actions Runners (runners)</summary>

This validation is required to limit the cost of high cost runners. See [Github Actions Runner Guidance](https://smartcontract-it.atlassian.net/l/cp/Rw0Gc08x).
For specific runner costs see "[What are the per-minute costs of the runners?](https://smartcontract-it.atlassian.net/wiki/spaces/RE/pages/861241466/Github+Actions+Runner+Guidance#What-are-the-per-minute-costs-of-the-runners%3F)"

##### \`runner-macos\`

* MacOS runners are very expensive in comparison to Ubuntu runners. If you need ARM64 architecture, consider using the new ARM64-based Ubuntu runners.
* If you need a MacOS runner, consider using the base runner especially if this is a public repository.
* If you must use an upgraded runner then see Ignoring Errors section below.

##### \`runner-ubuntu\`

* Per-minute compute costs scale with the number of cores. The base Ubuntu runner is the most cost-effective, especially for public repositories.
* If you must use an upgraded runner then see Ignoring Errors section below.

</details>

### Ignoring Errors

You can use the following string to ignore a line from validation \`ghv-ignore!\` (must be inlined).

When adding this it will trigger an error, this is expected. Please reach out to \`#team-releng\` for approval and bypass.

<details>
<summary>Examples</summary>

\`\`\`
runs-on: ubuntu-latest-99cores-999GB # ghv-ignore!

uses: org/action@sha # v1.0.0-node16 ghv-ignore!
\`\`\`
</details>
",
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "addSeparator": [MockFunction spy] {
          "calls": [
            [],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "addTable": [MockFunction spy] {
          "calls": [
            [
              [
                [
                  {
                    "data": "Filename",
                    "header": true,
                  },
                  {
                    "data": "Line Number",
                    "header": true,
                  },
                  {
                    "data": "Level",
                    "header": true,
                  },
                  {
                    "data": "Type",
                    "header": true,
                  },
                  {
                    "data": "Message",
                    "header": true,
                  },
                ],
              ],
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "write": [MockFunction spy] {
          "calls": [
            [],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`setSummary > should set summary with single file 1`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        [
          {
            "data": "Filename",
            "header": true,
          },
          {
            "data": "Line Number",
            "header": true,
          },
          {
            "data": "Level",
            "header": true,
          },
          {
            "data": "Type",
            "header": true,
          },
          {
            "data": "Message",
            "header": true,
          },
        ],
        [
          {
            "data": "foo.yml",
            "rowspan": "2",
          },
          {
            "data": "<a href="prefix.com/foo.yml#L1">1</a>",
            "rowspan": "1",
          },
          {
            "data": "error",
          },
          {
            "data": "sha-ref",
          },
          {
            "data": "sha ref error",
          },
        ],
        [
          {
            "data": "<a href="prefix.com/foo.yml#L3">3</a>",
            "rowspan": "1",
          },
          {
            "data": "warning",
          },
          {
            "data": "version-comment",
          },
          {
            "data": "version comment warning",
          },
        ],
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": {
        "addRaw": [MockFunction spy] {
          "calls": [
            [
              "


This CI step validates Github Actions workflows. If invoked by a pull request, it will only operate on the changed files.

It currently validates: Actions References, Actions Runner Types

### Fixing Errors

<details>
<summary>Action References (sha-ref, version-comment, node-version) </summary>

This validation is required to ensure that the action references use immutable SHAs, have a version comment, and are not using node16 or earlier.

The proper format for referencing a Github Action external to the repository is as follows:

\`<owner>/<repo>/<optional path>@<commit SHA> # <version tag relating to the SHA>\`

<details>
<summary>Examples</summary>

\`\`\`
actions/cache@ab5e6d0c87105b4c9c2047343972218f562e4319 # v4.0.1

smartcontractkit/chainlink-github-actions/github-app-token-issuer@5874ff7211cf5a5a2670bb010fbff914eaaae138 # v2.3.12
\`\`\`
</details>

##### <ref> is not a valid SHA

* Please reference a specific commit. This is because tags are mutable and pose a security risk
* Do not use things like \`@main\`, \`@branch/feature\`, \`@v4\`, or \`@v4.0.0\`.

##### No version comment found

* Ensure you have left a comment indicating the git tag (or version) associated with the SHA reference.
* \`@<commit> # v4.0.0\`

##### Action is using node...

* The action added is supposed to be run using a version that is not \`node20\`. This might create issues due to Github deprecating actions using \`node16\` and earlier.

</details>

<details>
<summary>Actions Runners (runners)</summary>

This validation is required to limit the cost of high cost runners. See [Github Actions Runner Guidance](https://smartcontract-it.atlassian.net/l/cp/Rw0Gc08x).
For specific runner costs see "[What are the per-minute costs of the runners?](https://smartcontract-it.atlassian.net/wiki/spaces/RE/pages/861241466/Github+Actions+Runner+Guidance#What-are-the-per-minute-costs-of-the-runners%3F)"

##### \`runner-macos\`

* MacOS runners are very expensive in comparison to Ubuntu runners. If you need ARM64 architecture, consider using the new ARM64-based Ubuntu runners.
* If you need a MacOS runner, consider using the base runner especially if this is a public repository.
* If you must use an upgraded runner then see Ignoring Errors section below.

##### \`runner-ubuntu\`

* Per-minute compute costs scale with the number of cores. The base Ubuntu runner is the most cost-effective, especially for public repositories.
* If you must use an upgraded runner then see Ignoring Errors section below.

</details>

### Ignoring Errors

You can use the following string to ignore a line from validation \`ghv-ignore!\` (must be inlined).

When adding this it will trigger an error, this is expected. Please reach out to \`#team-releng\` for approval and bypass.

<details>
<summary>Examples</summary>

\`\`\`
runs-on: ubuntu-latest-99cores-999GB # ghv-ignore!

uses: org/action@sha # v1.0.0-node16 ghv-ignore!
\`\`\`
</details>
",
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "addSeparator": [MockFunction spy] {
          "calls": [
            [],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "addTable": [MockFunction spy] {
          "calls": [
            [
              [
                [
                  {
                    "data": "Filename",
                    "header": true,
                  },
                  {
                    "data": "Line Number",
                    "header": true,
                  },
                  {
                    "data": "Level",
                    "header": true,
                  },
                  {
                    "data": "Type",
                    "header": true,
                  },
                  {
                    "data": "Message",
                    "header": true,
                  },
                ],
                [
                  {
                    "data": "foo.yml",
                    "rowspan": "2",
                  },
                  {
                    "data": "<a href="prefix.com/foo.yml#L1">1</a>",
                    "rowspan": "1",
                  },
                  {
                    "data": "error",
                  },
                  {
                    "data": "sha-ref",
                  },
                  {
                    "data": "sha ref error",
                  },
                ],
                [
                  {
                    "data": "<a href="prefix.com/foo.yml#L3">3</a>",
                    "rowspan": "1",
                  },
                  {
                    "data": "warning",
                  },
                  {
                    "data": "version-comment",
                  },
                  {
                    "data": "version comment warning",
                  },
                ],
              ],
            ],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
        "write": [MockFunction spy] {
          "calls": [
            [],
          ],
          "results": [
            {
              "type": "return",
              "value": [Circular],
            },
          ],
        },
      },
    },
  ],
}
`;
